Migrate Featured Items Blocks to TypeScript

sunyatasattva opened the pull request:
This PR migrates the `FeaturedItem` blocks to TypeScript. I tried to keep the runtime changes at a bare minimum. A few very minor of them were required. The biggest one was the refactoring of two pieces of code that were once again duplicates and also a bit hard to reason with (especially with TypeScript). These were refactored into [`with-update-button-attributes.tsx`](https://github.com/woocommerce/woocommerce-blocks/blob/9c34969f679e5cbf26159db3b3b80b3c6db94d73/assets/js/blocks/featured-items/with-update-button-attributes.tsx), where `withSelect` was also updated to use the more modern `useSelect` hook approach.

All the rest is adding the most type safety I could muster, and working around some upstream type limitations.

Fixes #6237, #6238

<!-- Don't forget to update the title with something descriptive. -->
<!-- If you can, add the appropriate labels -->

### Testing

#### Automated Tests
* [ ] Changes in this PR are covered by Automated Tests.
  * [ ] Unit tests
  * [ ] E2E tests

#### User Facing Testing

Only regression testing is needed for this PR. The changes introduced in this PR should be compared with a version without those changes and verified that no behavior has changed.

1. Add the **Featured Product** block to a page
2. Compare its functionality with the same block before these changes. Non-exhaustive list:
     - When added, it shows a full-width selection dropdown to choose the product
     - After choosing the product and clicking "Done", a loading spinner appears and the product is loaded
     - All the info are displayed correctly (default image, title, description, price, call to action)
     - The block toolbar shows the same tools (duotone, alignment, cropping/replacing, editing)
     - The block sidebar shows the same options and they all work correctly (colors, content, media settings, etc.); note that the “border” does not currently work.
     - The block appears resizable and resizing works the same way.
3. Verify the same for the **Featured Category** block

* [ ] Do not include in the Testing Notes <!-- Check this box if this PR can't be tested by users (ie: it doesn't include user-facing changes or it can't be tested without manually modifying the code). -->

### WooCommerce Visibility

<!-- Check this [this doc](../docs/blocks/feature-flags-and-experimental-interfaces.md) to see if the change is visible in WC core or not (part of the feature plugin or experimental)-->

* [x] WooCommerce Core
* [ ] Feature plugin
* [ ] Experimental

sunyatasattva committed:
Add typings to ConstrainedResizable

sunyatasattva committed:
Add types to  hook

sunyatasattva committed:
Add types to

sunyatasattva committed:
Change  to TypeScript constants

sunyatasattva committed:
Add types to

sunyatasattva committed:
Add types for

sunyatasattva committed:
Add types to `withEditMode` 

sunyatasattva committed:
Add types to `withEditingImage`

sunyatasattva committed:
Add types to `ImageEditor`

sunyatasattva committed:
Add types to block examples

sunyatasattva committed:
Add types for `BlockControls`

sunyatasattva committed:
Add types to `withFeaturedItem`

sunyatasattva committed:
Convert `block.js`s to TypeScript

Refactored some composed components into one inclusive HOC,
called `withUpdateButtonAttributes`.

sunyatasattva committed:
Add types for `inspector-controls`

sunyatasattva committed:
Fix TS errors

sunyatasattva committed:
Makes sure the item is defined in the inner block

sunyatasattva committed:
Merge branch 'trunk' into refactor/6237-6238-featured-items-typescript

frontdevde reviewed (commented):
When trying to follow the testing instructions and testing the Featured Product block, I immediately get an error in the editor upon block insertion. I double-checked and on trunk, appears to be working fine.

|trunk|branch|
|-|-|
|<img width="733" alt="Screenshot 2022-05-27 at 12 03 19" src="https://user-images.githubusercontent.com/1562646/170678725-0dc41564-dd6a-47d2-b1dd-8ca5560b899d.png">|<img width="787" alt="Screenshot 2022-05-27 at 12 04 40" src="https://user-images.githubusercontent.com/1562646/170678792-c463b937-30d7-4427-8144-d3fa8ac8383f.png">|

The error I'm seeing on this branch (and not on trunk):
<img width="513" alt="Screenshot 2022-05-27 at 12 00 28" src="https://user-images.githubusercontent.com/1562646/170680504-cac430ac-daa6-4375-8b50-cc328b2405bc.png">

frontdevde commented on assets/js/blocks/featured-items/featured-category/utils.ts:
```diff
@@ -1,15 +1,13 @@
-function getCategoryImageSrc( category ) {
+export function getCategoryImageSrc( category: WP_REST_API_Category ) {
 	if ( category && isObject( category.image ) ) {
 		return category.image.src;
```
Minor: `Property 'src' does not exist on type 'object'.ts(2339)`, same for image.id below.

sunyatasattva replied:
Good catch! 👏

frontdevde commented on assets/js/blocks/featured-items/with-edit-mode.tsx:
```diff
@@ -40,9 +73,15 @@ export const withEditMode = ( { description, editLabel, icon, label } ) => (
 				{ description }
 				<div className={ `${ className }__selection` }>
 					{ name === BLOCK_NAMES.featuredCategory && (
+						// Ignoring this TS error for now as it seems that `ProductCategoryControl`
```
If we haven't already we could create an issue to revisit `ProductCategoryControl` types. My worry would be that otherwise the ts-ignore might just stick around forever.

sunyatasattva replied:
True. Honestly I hadn't added it because my worry was that instead the ticket would stick around forever, or be lost in the stale oblivion. But you are right.

frontdevde commented on assets/js/blocks/featured-items/with-edit-mode.tsx:
```diff
@@ -56,9 +95,20 @@ export const withEditMode = ( { description, editLabel, icon, label } ) => (
+									? [ attributes.productId ]
+									: [ 0 ]
+							}
+							// `ProductControl` is not yet a TypeScript file and the types
```
Same as above, probably best to create an issue so we keep track of this.

frontdevde commented on assets/js/blocks/featured-items/with-featured-item.tsx:
```diff
@@ -3,27 +3,103 @@
+		| FeaturedProductRequiredAttributes
+	 ) &
+		EditorBlock< T >[ 'attributes' ] & {
+			// This is hardcoded because border is not yet included in Gutenberg's
```
If we haven't yet, it might make sense to log this as an issue and/or contribute upstream.

sunyatasattva replied:
Yep, a few contributions upstream are on my notes.

sunyatasattva committed:
Fix category utility functions types

sunyatasattva committed:
Add todo comment regarding adding types

sunyatasattva committed:
Add another todo comment

sunyatasattva committed:
Merge commit 'e650560bee90732b76ebc4b14255c5ab74c9277b' into refactor/6237-6238-featured-items-typescript

sunyatasattva committed:
Fix isImageObject typeguard

sunyatasattva committed:
Fix edit mode not rendering on first render

The HOC `withTransformSingleSelectToMultipleSelect` allows for the wrong
type to be passed to its wrapped component: if a single element is
passed, that is wrapped into an array.

This didn't play well with TypeScript which obviously complains about
the wrong type being passed to the component.

So we now allow for empty arrays to be passed and parsed as "nil".

sunyatasattva commented:
@frontdevde @albarin Feedback addressed!

frontdevde commented:
I'm still seeing the same block error (see above) when I'm trying to insert the Featured Product block on this branch (not on trunk). Can you reproduce that issue?

sunyatasattva committed:
Fix Featured Product block bugging out when product is selected

frontdevde reviewed (approved):
Thank you for this valiant effort and for addressing the feedback!
When retesting with the most recent changes functionally everything appears to be working as expected. Let's get this in as the improvement over the status quo it is and iterate on anything else from there.
